plugins/inputs/zfs/zfs_freebsd.go

```diff
@@ -87,6 +87,61 @@ func (z *Zfs) gatherPoolStats(acc telegraf.Accumulator) (string, error) {
 	return strings.Join(pools, "::"), nil
 }
 
+func (z *Zfs) gatherDatasetStats(acc telegraf.Accumulator) (string, error) {
+
+	lines, err := z.zdataset()
+	if err != nil {
+		return "", err
+	}
+
+	datasets := []string{}
+	for _, line := range lines {
+		col := strings.Split(line, "\t")
+
+		datasets = append(datasets, col[0])
+	}
+
+	if z.DatasetMetrics {
+		for _, line := range lines {
+			col := strings.Split(line, "\t")
+			if len(col) != 5 {
+				continue
+			}
+
+			tags := map[string]string{"dataset": col[0]}
+			fields := map[string]interface{}{}
+
+			avail, err := strconv.ParseInt(col[1], 10, 64)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing avail: %s", err)
+			}
+			fields["avail"] = avail
+
+			used, err := strconv.ParseInt(col[2], 10, 64)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing used: %s", err)
+			}
+			fields["used"] = used
+
+			usedsnap, err := strconv.ParseInt(col[3], 10, 64)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing usedsnap: %s", err)
+			}
+			fields["usedsnap"] = usedsnap
+
+			usedds, err := strconv.ParseInt(col[4], 10, 64)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing usedds: %s", err)
+			}
+			fields["usedds"] = usedds
```

Comment:
This is a re-occurring pattern. Assuming that all returned properties (except for the name) are integers we can do
```suggestion
		for i, key := range properties[1:] {
			value, err := strconv.ParseInt(col[i+1], 10, 64)
			if err != nil {
				return "", fmt.Errorf("Error parsing %s: %s", key, err)
			}
			fields[key] = value
		}
```